dask_expr/shuffle.py

```diff
@@ -609,3 +623,67 @@ def operation(cls, df, index, name: str, npartitions: int):
         else:
             index = partitioning_index(index, npartitions)
         return df.assign(**{name: index})
+
+
+class SetIndex(Expr):
+    _parameters = ["frame", "_other", "drop", "sorted", "user_divisions"]
+    _defaults = {"drop": True, "sorted": False, "user_divisions": None}
+
+    def _divisions(self):
+        if self.user_divisions is not None:
+            return self.user_divisions
+        return self._calculate_divisions()
+
+    @property
+    def _meta(self):
+        if isinstance(self._other, Expr):
+            other = self._other._meta
+        else:
+            other = self._other
+        return self.frame._meta.set_index(other, drop=self.drop)
+
+    @property
+    def other(self):
+        if isinstance(self._other, Expr):
+            return self._other
+        return self.frame[self._other]
+
+    @functools.lru_cache  # noqa: B019
```

Comment:
Are we ok with caching here?

Reply:
Fine with me.  Thoughts on merging this method with `_divisions` above?  Will this ever be used separately?

Reply:
We will have to compute more statistics ("presorted" for example") in this method when adding support for other keywords and code-paths, which will make this method different from ``_divisions``.

Reply:
I'm not sure I understand, will there be times when we call this method outside of `_divisions`

Reply:
Yes!

Reply:
Then happy to retract my comment.  Thanks for clarifying.